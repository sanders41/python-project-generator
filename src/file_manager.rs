use std::{fs::File, io::prelude::*, path::PathBuf};

use anyhow::Result;

use crate::project_info::ProjectInfo;

pub fn save_file_with_content(file_path: &PathBuf, file_content: &str) -> Result<()> {
    let mut file = File::create(file_path)?;
    file.write_all(file_content.as_bytes())?;

    Ok(())
}

pub fn save_empty_src_file(project_info: &ProjectInfo, file_name: &str) -> Result<()> {
    let module = project_info.source_dir.replace([' ', '-'], "_");
    let file_path = project_info
        .base_dir()
        .join(format!("{}/{}", &module, file_name));
    File::create(file_path)?;

    Ok(())
}
