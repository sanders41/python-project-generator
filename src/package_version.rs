use std::fmt;
use std::time::Duration;

use anyhow::Result;

#[derive(Debug, PartialEq, Eq)]
pub enum PythonPackage {
    Maturin,
    MyPy,
    PreCommit,
    Pytest,
    PytestCov,
    Ruff,
    Tomli,
}

impl fmt::Display for PythonPackage {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            PythonPackage::Maturin => write!(f, "maturin"),
            PythonPackage::MyPy => write!(f, "mypy"),
            PythonPackage::PreCommit => write!(f, "pre-commit"),
            PythonPackage::Pytest => write!(f, "pytest"),
            PythonPackage::PytestCov => write!(f, "pytest-cov"),
            PythonPackage::Ruff => write!(f, "ruff"),
            PythonPackage::Tomli => write!(f, "tomli"),
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum PreCommitHook {
    PreCommit,
    MyPy,
    Ruff,
}

impl fmt::Display for PreCommitHook {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            PreCommitHook::MyPy => write!(f, "mypy"),
            PreCommitHook::PreCommit => write!(f, "pre-commit"),
            PreCommitHook::Ruff => write!(f, "ruff"),
        }
    }
}

pub trait LatestVersion {
    fn get_latest_version(&mut self) -> Result<()>;
}

#[derive(Debug)]
pub struct PreCommitHookVersion {
    pub hook: PreCommitHook,
    pub repo: String,
    pub rev: String,
}

impl LatestVersion for PreCommitHookVersion {
    fn get_latest_version(&mut self) -> Result<()> {
        let client = reqwest::blocking::Client::new();
        let api_url = format!(
            "{}/releases",
            self.repo
                .replace("https://github.com", "https://api.github.com/repos")
        );
        let response = client
            .get(api_url)
            .header(reqwest::header::USER_AGENT, "python-project-generator")
            .timeout(Duration::new(5, 0))
            .send()?
            .text()?;
        let info: Vec<serde_json::Value> = serde_json::from_str(&response)?;
        for i in info {
            if i["draft"] == false && i["prerelease"] == false {
                self.rev = i["tag_name"].to_string().replace('"', "");
                break;
            }
        }

        Ok(())
    }
}

impl PreCommitHookVersion {
    pub fn new(hook: PreCommitHook) -> Self {
        let rev = default_pre_commit_rev(&hook);
        let repo = pre_commit_repo(&hook);
        PreCommitHookVersion { hook, repo, rev }
    }
}

#[derive(Debug)]
pub struct PythonPackageVersion {
    pub package: PythonPackage,
    pub version: String,
}

impl LatestVersion for PythonPackageVersion {
    fn get_latest_version(&mut self) -> Result<()> {
        let name = self.package.to_string();
        let url = format!("https://pypi.org/pypi/{}/json", name);
        let client = reqwest::blocking::Client::new();
        let response = client
            .get(url)
            .timeout(Duration::new(5, 0))
            .send()?
            .text()?;
        let info: serde_json::Value = serde_json::from_str(&response)?;
        self.version = info["info"]["version"].to_string().replace('"', "");

        Ok(())
    }
}

impl PythonPackageVersion {
    pub fn new(package: PythonPackage) -> Self {
        let version = default_version(&package);

        PythonPackageVersion { package, version }
    }
}

#[derive(Debug)]
pub struct RustPackageVersion {
    pub name: String,
    pub version: String,
    pub features: Option<Vec<String>>,
}

impl LatestVersion for RustPackageVersion {
    fn get_latest_version(&mut self) -> Result<()> {
        let url = format!("https://crates.io/api/v1/crates/{}", self.name);
        let client = reqwest::blocking::Client::new();
        let response = client
            .get(url)
            .header(reqwest::header::USER_AGENT, "python-project-generator")
            .timeout(Duration::new(5, 0))
            .send()?
            .text()?;
        let info: serde_json::Value = serde_json::from_str(&response)?;
        self.name = info["crate"]["id"].to_string().replace('"', "");
        self.version = info["crate"]["max_stable_version"]
            .to_string()
            .replace('"', "");

        Ok(())
    }
}

pub fn default_version(package: &PythonPackage) -> String {
    match package {
        PythonPackage::Maturin => "1.7.1".to_string(),
        PythonPackage::MyPy => "1.11.2".to_string(),
        PythonPackage::PreCommit => "3.8.0".to_string(),
        PythonPackage::Pytest => "8.3.2".to_string(),
        PythonPackage::PytestCov => "5.0.0".to_string(),
        PythonPackage::Ruff => "0.6.3".to_string(),
        PythonPackage::Tomli => "2.0.1".to_string(),
    }
}

pub fn default_pre_commit_rev(hook: &PreCommitHook) -> String {
    match hook {
        PreCommitHook::MyPy => "v1.11.2".to_string(),
        PreCommitHook::PreCommit => "v4.6.0".to_string(),
        PreCommitHook::Ruff => "v0.6.3".to_string(),
    }
}

pub fn pre_commit_repo(hook: &PreCommitHook) -> String {
    match hook {
        PreCommitHook::MyPy => "https://github.com/pre-commit/mirrors-mypy".to_string(),
        PreCommitHook::PreCommit => "https://github.com/pre-commit/pre-commit-hooks".to_string(),
        PreCommitHook::Ruff => "https://github.com/astral-sh/ruff-pre-commit".to_string(),
    }
}
