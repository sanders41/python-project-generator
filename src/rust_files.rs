use anyhow::Result;
use colored::*;
use rayon::prelude::*;

use crate::file_manager::save_file_with_content;
use crate::licenses::license_str;
use crate::package_version::{LatestVersion, RustPackageVersion};
use crate::project_info::{LicenseType, ProjectInfo};

fn build_latest_dependencies(download_latest_packages: bool) -> String {
    let mut version_string = String::new();
    let mut packages = vec![RustPackageVersion {
        name: "pyo3".to_string(),
        version: "0.22.2".to_string(),
        features: Some(vec!["extension-module".to_string()]),
    }];

    if download_latest_packages {
        packages.par_iter_mut().for_each(|package| {
            if package.get_latest_version().is_err() {
                let error_message = format!(
                    "Error retrieving latest crate version for {}. Using default.",
                    package.name
                );
                println!("\n{}", error_message.yellow());
            }
        })
    }

    for package in packages {
        if let Some(features) = &package.features {
            let mut feature_str = "[".to_string();
            for feature in features {
                feature_str.push_str(&format!(r#""{feature}", "#));
            }

            feature_str.truncate(feature_str.len() - 2);
            feature_str.push(']');

            version_string.push_str(&format!(
                "{} = {{ version = \"{}\", features = {} }}\n",
                package.name, package.version, feature_str
            ));
        } else {
            version_string.push_str(&format!(
                "{} = {{ version = \"{}\" }}\n",
                package.name, package.version
            ));
        }
    }

    version_string.trim().to_string()
}

fn create_cargo_toml_file(
    project_slug: &str,
    project_description: &str,
    source_dir: &str,
    license_type: &LicenseType,
    download_latest_packages: bool,
) -> String {
    let versions = build_latest_dependencies(download_latest_packages);
    let license = license_str(license_type);
    let name = source_dir.replace([' ', '-'], "_");

    format!(
        r#"[package]
name = "{project_slug}"
version = "0.1.0"
description = "{project_description}"
edition = "2021"
license = "{license}"
readme = "README.md"

[lib]
name = "_{name}"
crate-type = ["cdylib"]

[dependencies]
{versions}
"#
    )
}

pub fn save_cargo_toml_file(project_info: &ProjectInfo) -> Result<()> {
    let file_path = project_info.base_dir().join("Cargo.toml");
    let content = create_cargo_toml_file(
        &project_info.project_slug,
        &project_info.project_description,
        &project_info.source_dir,
        &project_info.license,
        project_info.download_latest_packages,
    );

    save_file_with_content(&file_path, &content)?;

    Ok(())
}

fn create_lib_file(source_dir: &str) -> String {
    let module = source_dir.replace([' ', '-'], "_");
    format!(
        r#"use pyo3::prelude::*;

#[pyfunction]
fn sum_as_string(a: usize, b: usize) -> PyResult<String> {{
    Ok((a + b).to_string())
}}

#[pymodule]
fn _{module}(_py: Python, m: &Bound<'_, PyModule>) -> PyResult<()> {{
    m.add_function(wrap_pyfunction!(sum_as_string, m)?)?;
    Ok(())
}}
"#
    )
}

pub fn save_lib_file(project_info: &ProjectInfo) -> Result<()> {
    let file_path = project_info.base_dir().join("src/lib.rs");
    let content = create_lib_file(&project_info.source_dir);

    save_file_with_content(&file_path, &content)?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::project_info::{LicenseType, ProjectInfo, ProjectManager};
    use insta::assert_yaml_snapshot;
    use std::fs::create_dir_all;
    use tempfile::tempdir;

    fn project_info_dummy() -> ProjectInfo {
        ProjectInfo {
            project_name: "My project".to_string(),
            project_slug: "my-project".to_string(),
            source_dir: "my_project".to_string(),
            project_description: "This is a test".to_string(),
            creator: "Arthur Dent".to_string(),
            creator_email: "authur@heartofgold.com".to_string(),
            license: LicenseType::Mit,
            copyright_year: Some("2023".to_string()),
            version: "0.1.0".to_string(),
            python_version: "3.12".to_string(),
            min_python_version: "3.9".to_string(),
            project_manager: ProjectManager::Maturin,
            is_application: true,
            github_actions_python_test_versions: vec![
                "3.9".to_string(),
                "3.10".to_string(),
                "3.11".to_string(),
                "3.12".to_string(),
            ],
            max_line_length: 100,
            use_dependabot: true,
            dependabot_schedule: None,
            dependabot_day: None,
            use_continuous_deployment: true,
            use_release_drafter: true,
            use_multi_os_ci: true,
            download_latest_packages: false,
            project_root_dir: Some(tempdir().unwrap().path().to_path_buf()),
        }
    }

    #[test]
    fn test_save_cargo_toml_file() {
        let project_info = project_info_dummy();
        let base = project_info.base_dir();
        create_dir_all(base.join(&project_info.project_slug)).unwrap();
        let expected_file = base.join("Cargo.toml");
        save_cargo_toml_file(&project_info).unwrap();

        assert!(expected_file.is_file());

        let content = std::fs::read_to_string(expected_file).unwrap();

        assert_yaml_snapshot!(content);
    }

    #[test]
    fn test_save_lib_file() {
        let project_info = project_info_dummy();
        let base = project_info.base_dir();
        create_dir_all(base.join("src")).unwrap();
        let expected_file = base.join("src/lib.rs");
        save_lib_file(&project_info).unwrap();

        assert!(expected_file.is_file());

        let content = std::fs::read_to_string(expected_file).unwrap();

        assert_yaml_snapshot!(content);
    }
}
