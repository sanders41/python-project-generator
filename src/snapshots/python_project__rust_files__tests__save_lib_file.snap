---
source: src/rust_files.rs
expression: content
---
"use pyo3::prelude::*;\n\n#[pyfunction]\nfn sum_as_string(a: usize, b: usize) -> PyResult<String> {\n    Ok((a + b).to_string())\n}\n\n#[pymodule]\nfn _my_project(_py: Python, m: &Bound<'_, PyModule>) -> PyResult<()> {\n    m.add_function(wrap_pyfunction!(sum_as_string, m)?)?;\n    m.add(\"__version__\", env!(\"CARGO_PKG_VERSION\"))?;\n    Ok(())\n}\n"
